bundles/org.openhab.binding.alarmdecoder/src/main/java/org/openhab/binding/alarmdecoder/internal/protocol/LRRMessage.java

```diff
@@ -0,0 +1,73 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.alarmdecoder.internal.protocol;
+
+import java.util.List;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+
+/**
+ * The {@link LRRMessage} class represents a parsed LRR message.
+ * Based partly on code from the OH1 alarmdecoder binding by Bernd Pfrommer and Lucky Mallari.
+ *
+ * @author Bob Adair - Initial contribution
+ */
+@NonNullByDefault
+public class LRRMessage extends ADMessage {
+
+    // Example: !LRR:012,1,CID_1441,ff
+
+    /** Event data contains user number or zone number for the event */
+    public final String eventData;
+
+    /** Partition event applies to. 0 means all partitions. */
+    public final int partition;
+
+    /** CID message for event as defined in SIA DC-05-1999.09 standard */
+    public final String cidMessage;
+
+    /** Report code */
+    public final String reportCode;
+
+    public LRRMessage(String message) throws IllegalArgumentException {
+        super(message);
+
+        String topLevel[] = message.split(":");
+        if (topLevel.length != 2) {
+            throw new IllegalArgumentException("multiple colons in LRR message");
+        }
+
+        List<String> parts = splitMsg(topLevel[1]);
+
+        // Apparently the 4th part of the LRR message may not be included depending on version
+        if (parts.size() < 3 || parts.size() > 4) {
+            throw new IllegalArgumentException("Invalid number of parts in LRR message");
+        }
+
+        eventData = parts.get(0);
+        cidMessage = parts.get(2);
+        reportCode = parts.size() == 4 ? parts.get(3) : "";
+
+        try {
+            int p = 0;
+            try {
+                p = Integer.parseInt(parts.get(1));
+            } catch (NumberFormatException e) {
+                p = Integer.parseInt(parts.get(1), 16);
```

Comment:
How would this is correct if you are given a hexadecimal number that doesn't have any letters in it?

Reply:
Looking through the docs again, I think this field is unlikely to ever be a hex number. However, if it were, I believe it would have a 0x prefix. There is equivalent code in the keypad message parser, which comes from the V1 binding.

Reply:
I could not find precisely what the protocol specs say about that field. The specs don't even say it's an integer, but the example they are giving is:
"Event Data:
This is either the user number of the person performing the action or the zone that was bypassed."

http://www.alarmdecoder.com/wiki/index.php/Protocol#Format_5

I don't know, might be safest to treat it as a string? (Sorry, I don't receive LRR messages on my system, so I don't have any example messages).

Reply:
Thanks @berndpfrommer. Yes, the AD protocol spec wasn't too much help. From looking at the SIA DC-05-1999.09 document (pp. 5-8) it seems clear that zone and user are both integers, but there is no clear indication as to what number system the AD would use to send them other than the example in there docs which appears to be decimal. Maybe @LuckyMallari would have some input on this?

Using a string could work as well if we can't come up with a definitive answer.  BTW - I've been planning to add another channel to the lrr thing that provides the whole LRR message payload as a string for better backward compatibility with the V1 binding, but I thought I would do that after this initial submission was merged.

Reply:
> Thanks @berndpfrommer. Yes, the AD protocol spec wasn't too much help. From looking at the SIA DC-05-1999.09 document (pp. 5-8) it seems clear that zone and user are both integers, but there is no clear indication as to what number system the AD would use to send them other than the example in there docs which appears to be decimal. Maybe @LuckyMallari would have some input on this?
> 
> Using a string could work as well if we can't come up with a definitive answer. BTW - I've been planning to add another channel to the lrr thing that provides the whole LRR message payload as a string for better backward compatibility with the V1 binding, but I thought I would do that after this initial submission was merged.

Sorry for being inactive on the thread.. busy with stuff ;) Had to dig from my old posts, but here's some info:

https://www.openhab.org/addons/bindings/alarmdecoder1/
```
LRR format from Alarm Decoder is:

!LRR:<data>,<partition>,<event>
```

and
```
Because <data> could be anything, it cannot be tied to a particular item as opposed to item bindings such as Contacts mentioned above. Instead, the entire message is sent to the binding.

!LRR:<entiremessage>
```

Now, if you look at the [history](http://www.alarmdecoder.com/wiki/index.php?title=LRR_Support&action=history) story of the page, it has not been updated since 2013, so my comments are still correct.

I am still using the AD binding v1, and still works flawlessly at the time of this writing. Here are some same LRR messages captured from my Vista20P:

```
!SER2SOCK Connected
!SER2SOCK SERIAL_CONNECTED
##1
!Sending.done
[1001000100000000----],008,[f70000ff1008001c08020000000000],"THE MALLARI'S     R                                                                             eady to Arm  "
#1
!Sending.done
!RFX:0746846,80
#
!Sending.done
1
!RFX:0746846,00
!Sending.done
[1001000100000000----],008,[f70000ff1008001c08020000000000],"THE MALLARI'S     R                                                                             eady to Arm  "
#2
!Sending.done
[0101050100000000----],008,[f70000ff1008050c0c020000000000],"ARMED ***AWAY***You                                                                              may exit now"
!LRR:000,1,ARM_AWAY
!Sending.done
!Sending.done
!Sending.done
!Sending.done
!Sending.done
[1001010100000000----],008,[f70000ff1008011c08020000000000],"THE MALLARI'S     R                                                                             eady to Arm  "
!RFX:0322328,84
[1001000100000000----],008,[f70000ff1008001c08020000000000],"THE MALLARI'S     R                                                                             eady to Arm  "
!LRR:003,1,OPEN
!Sending.done
!Sending.done
!LRR:000,1,ARM_STAY
[0011030100000000----],008,[f70000ff1008038c08020000000000],"ARMED ***STAY***You                                                                              may exit now"
!Sending.done
!Sending.done
!Sending.done
[0011000100000000----],008,[f70000ff1008008c08020000000000],"ARMED ***STAY***You                                                                              may exit now"
!Sending.done
!Sending.done
[1001010100000000----],008,[f70000ff1008011c08020000000000],"THE MALLARI'S     R                                                                             eady to Arm  "
[1001000100000000----],008,[f70000ff1008001c08020000000000],"THE MALLARI'S     R                                                                             eady to Arm  "
!LRR:003,1,OPEN
!Sending.done
!Sending.done
!Sending.done
!Sending.done
!Sending.done
[0011030100000000----],008,[f70000ff1008038c08020000000000],"ARMED ***STAY***You                                                                              may exit now"
!LRR:003,1,ARM_STAY
[0011000100000000----],008,[f70000ff1008008c08020000000000],"ARMED ***STAY***You                                                                              may exit now"
!Sending.done
!Sending.done
!Sending.done
!Sending.done
!Sending.done
[0011000100000000----],008,[f70000ff1008008c08020000000000],"ARMED ***STAY***You                                                                              may exit now"
!Sending.done
!Sending.done
!Sending.done
!Sending.done
!Sending.done
[1001010100000000----],008,[f70000ff1008011c08020000000000],"THE MALLARI'S     R                                                                             eady to Arm  "
!Sending.done
!Sending.done
!Sending.done
!Sending.done
[1001000100000000----],008,[f70000ff1008001c08020000000000],"THE MALLARI'S     R                                                                             eady to Arm  "
!LRR:002,1,OPEN
!Sending.done
!LRR:002,1,ARM_STAY
[0011030100000000----],008,[f70000ff1008038c08020000000000],"ARMED ***STAY***You                                                                              may exit now"
!Sending.done
!Sending.done
!Sending.done
!Sending.done
!Sending.done
!RFX:0702437,84
[0011000100000000----],008,[f70000ff1008008c08020000000000],"ARMED ***STAY***You may exit now"
!Sending.done
!Sending.done
!Sending..done
!RFX:0638432,80
!Sending.done
[1001010100000000----],008,[f70000ff1008011c08020000000000],"THE MALLARI'S     Ready to Arm  "
!RFX:0638432,00
[1001000100000000----],008,[f70000ff1008001c08020000000000],"THE MALLARI'S     Ready to Arm  "
!LRR:002,1,OPEN
```

Here's how to interpret the LRR messages:
* `!LRR:000,1,ARM_AWAY` - I simply sent # 2 (which is same as hitting # 2 on keypad. This means the system was Armed (away), but it is `000` because this is partition #. In my OH rules, I don't treat this as partition, on ARM/DISARM states, I treat `000` as "system" user, which means an open/arm was invoked without using a PIN (like doing a # 2) from keypad. 
* `!LRR:002,1,OPEN` - Disarmed by User 2
* `!LRR:003,1,ARM_STAY` - Armed stay by user 3. This is when you hit PIN then Stay (in Vista20P keypads, this is 3). For example, if user 3 pin code is 1234, then you get this by hitting 12343 on keypad.
* `!LRR:00,1,OPEN` - Disarmed by User 3


LRR Event Types can be retrieved [here](http://www.alarmdecoder.com/wiki/index.php/LRR_Support)

Note that LRR events will need to be enabled by user on the alarm itself. For example, for OPEN to show up in LRR, OPEN reports should be enabled (field *65)

You need to
* Filter (or switch/case) events such as OPEN/ARM_AWAY/ARM_STAY and treat the field as USER 
* Events such as LOWBAT/TROUBLE/etc and treat them as partition

I just dont have enough free time to test which events are user events and which are system/partition events

But to quickly answer your question, they are INTEGERS

Reply:
Thanks @LuckyMallari ! That is great info. So we can assume that *data* and *partition* will always be decimal integers. I've removed the hex parsing code that @cpmeister originally commented on, and also added your LRR messages to the test data I'm using.

Do you know what determines whether the LRR messages will be of the format in your data or the "CID" format mentioned in the current protocol docs (e.g. "!LRR:012,1,CID_1441,ff")? The parsing code will work with either, so I only ask out of curiosity, and maybe for purposes of documentation. I see there is an AD firmware capability called "L2 - Long range radio messages (!LRR) with CID reporting format", so I've been assuming it is related to firmware version and/or configuration.

Reply:
I thought I read someplace that LRR is always in Ademco Contact ID format (CID) regardless of the alarm setting for the phoneline. 

The LRR seems to be another destination for reports. What is the benefit of LRR support in the binding if the same messages are sent to the keypad?